sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala

```diff
@@ -160,34 +161,46 @@ class CacheManager extends Logging {
   /**
    * Tries to re-cache all the cache entries that refer to the given plan.
    */
-  def recacheByPlan(spark: SparkSession, plan: LogicalPlan): Unit = writeLock {
+  def recacheByPlan(spark: SparkSession, plan: LogicalPlan): Unit = readLock {
     recacheByCondition(spark, _.find(_.sameResult(plan)).isDefined)
   }
 
   private def recacheByCondition(
       spark: SparkSession,
       condition: LogicalPlan => Boolean,
       clearCache: Boolean = true): Unit = {
-    val it = cachedData.iterator()
     val needToRecache = scala.collection.mutable.ArrayBuffer.empty[CachedData]
-    while (it.hasNext) {
-      val cd = it.next()
-      if (condition(cd.plan)) {
-        if (clearCache) {
-          cd.cachedRepresentation.cacheBuilder.clearCache()
+    writeLock {
+      val it = cachedData.iterator()
+      while (it.hasNext) {
+        val cd = it.next()
+        if (condition(cd.plan)) {
+          if (clearCache) {
+            cd.cachedRepresentation.cacheBuilder.clearCache()
```

Comment:
same here, we can put plans to recache in a list first, and recache them without lock.

Reply:
Again, this only works if it is safe to call clearCache() while the item is still in the cache.  I'm not familiar with this code enough to 100% know that it's ok, but made the change anyway here.

Reply:
I refactored and I think it makes more sense now.